pennylane_lightning_gpu/src/algorithms/AdjointDiffGPU.hpp

```diff
@@ -208,28 +226,48 @@ template <class T = double> class AdjointJacobianGPU {
      * @brief Utility method to update the Jacobian at a given index by
      * calculating the overlap between two given states.
      *
-     * @param sv1 Statevector <sv1|. Data will be conjugated.
+     * @param sv1s vector of statevector <sv1| (one for each observable). Data
+     * will be conjugated.
      * @param sv2 Statevector |sv2>
      * @param jac Jacobian receiving the values.
      * @param scaling_coeff Generator coefficient for given gate derivative.
-     * @param obs_index ObservableGPU index position of Jacobian to update.
+     * @param num_observables the number of observables of Jacobian to update.
      * @param param_index Parameter index position of Jacobian to update.
+     * @param device_buffer_jac_single_param a workspace buffer on the device of
+     * size num_observables elements. Data will be ignored and overwritten.
+     * @param host_buffer_jac_single_param a workspace buffer on the host of
+     * size num_observables elements. Data will be ignored and overwritten.
      */
-    inline void updateJacobian(const StateVectorCudaManaged<T> &sv1,
-                               const StateVectorCudaManaged<T> &sv2,
-                               std::vector<std::vector<T>> &jac,
-                               T scaling_coeff, size_t obs_index,
-                               size_t param_index) {
-        PL_ABORT_IF_NOT(sv1.getDataBuffer().getDevTag().getDeviceID() ==
-                            sv2.getDataBuffer().getDevTag().getDeviceID(),
-                        "Data exists on different GPUs. Aborting.");
-
-        jac[obs_index][param_index] =
-            -2 * scaling_coeff *
-            innerProdC_CUDA(sv1.getData(), sv2.getData(), sv1.getLength(),
-                            sv1.getDataBuffer().getDevTag().getDeviceID(),
-                            sv1.getDataBuffer().getDevTag().getStreamID())
-                .y;
+    inline void
+    updateJacobian(const std::vector<StateVectorCudaManaged<T>> &sv1s,
+                   const StateVectorCudaManaged<T> &sv2,
+                   std::vector<std::vector<T>> &jac, T scaling_coeff,
+                   size_t num_observables, size_t param_index,
+                   DataBuffer<CFP_t, int> &device_buffer_jac_single_param,
+                   thrust::host_vector<CFP_t> &host_buffer_jac_single_param) {
+        host_buffer_jac_single_param.clear();
```

Comment:
I added a `vector.clear();` in the begining and the end of the function to clarify that neither the input nor the output of this buffer will be used.
For two consequitive `clear()`, at least one of them is essentially a no-op. If we remove one, I would also remove the other because it's not strictly necessary and then we would save a `num_observables` assignments of vector elements on `resize(num_observables)`.

Reply:
Sure, no problem then to leave as is, or remove both. Whichever you prefer.

Reply:
No strong preference for either - I left it for now.